x/ecocredit/module/module.go

```diff
@@ -3,27 +3,33 @@ package module
 import (
 	"encoding/json"
 
-	"github.com/gorilla/mux"
-	"github.com/grpc-ecosystem/grpc-gateway/runtime"
-	"github.com/spf13/cobra"
-
-	abci "github.com/tendermint/tendermint/abci/types"
-
 	sdkclient "github.com/cosmos/cosmos-sdk/client"
 	"github.com/cosmos/cosmos-sdk/codec"
 	codectypes "github.com/cosmos/cosmos-sdk/codec/types"
 	sdk "github.com/cosmos/cosmos-sdk/types"
 	"github.com/cosmos/cosmos-sdk/types/module"
+	"github.com/gorilla/mux"
+	"github.com/grpc-ecosystem/grpc-gateway/runtime"
+	"github.com/spf13/cobra"
+	abci "github.com/tendermint/tendermint/abci/types"
 
 	"github.com/regen-network/regen-ledger/x/ecocredit"
 	"github.com/regen-network/regen-ledger/x/ecocredit/server"
 )
 
 type AppModule struct {
 	Key sdk.StoreKey
+	Srv server.Server
```

Comment:
What value is there in adding this here? Where is it going to need to be publicly used?

Reply:
I wrote about it on [Discord]() and in the commit message: https://github.com/regen-network/regen-ledger/pull/153/commits/ff2a9036c68ee5fd9b96d0b2a9dcfe46c8e392f9

The whole idea is to allow modules to make it dependency: other modules can mock a server if needed.

Reply:
In fact the main use-case was to use them in handlers. Since we don't need to implement a Router, then it has less utility -  app module doesn't need to hold the server instance any more (at least I don't see other methods where it's needed, `RegisterGRPCGatewayRoutes` also doesn't use the server).

The only reason I to keep it here is to be able to customize the server within the `AppModule.RegisterServices`. Since the AppModule has much less utility now, then probably this is not needed. We can parametrize the package `ecocredits/RegisterServices` method (note, it's not in the server anymore, and it will take Server as an argument) and  external tests can use that method instead of AppModule. What do you think?

Reply:
Right, the app module doesn't need a server instance at all with this design and that's intentional. Maybe I need to document the reasons more but here's my thinking:
1. other modules shouldn't hold "keepers"/servers directly (according to the WIP ADR 033), instead they use clients which handle permissions correctly. So a direct server instance is undesirable because it breaks ocaps/encapsulation
2. an `AppModule` can eventually become sort of a "module descriptor" which instantiates module services but doesn't maintain any of its own state. Module descriptors would _only_ contain config parameters. In `RegisterServices`, services are constructed and registered but not owned by the app module - rather ownership is transfered to the various routers/service managers.

 My preference right now is to leave `RegisterServices` in `server` as it was. I originally had things the way you have them here, but I changed it because that felt like it had better encapsulation. Basically with `RegisterServices` in `server`, tests don't need to know how services are getting implemented or ever have a concrete server implementation - all of that is hidden so there is proper ocaps provided by the clients.

Does that make sense?

In general I'm leaning in the direction of AppModule being much leaner and tests would mainly need `RegisterInterfaces` and `RegisterServices` instead of all the other AppModule stuff currently there.

Reply:
I'm thinking that a better, less high coupled approach might be something like:
- what is `types` in the SDK and now the root in this approach just has types and maybe something like a `ModuleDescriptor` with just `RegisterInterfaces`
- `server` has a lightweight `ServerModule` which just has `RegisterServices` + `RegisterInterfaces`
- there are separate client modules wired up separately like `CLIClientModule`, `GRPCGatewayClientModule`, etc.

This has a dependency structure where `x/{module}/server` and `x/{module}/client` depend on `x/{module}` and app.go only imports `x/{module}/server.ServerModule`, and we don't have everything like clients and simulations getting imported into app.go.

Reply:
This makes sense. I like that. Although I would keep the `AppModule` name - it describes it's goal very well. `x/{module}` shouldn't import `x/{module}/subpackage`.

Reply:
This makes sense to me too. Some comments:
- why not call it `ModuleServer` instead of `ServerModule`: it's literally a module's server.
- I'm not sure why we have `RegisterInterfaces` in both `ServerModule` and `ModuleDescriptor`. Only once should be enough? My preference goes to the root package.
- While `CLIClientModule` and `GRPCGatewayClientModule` makes things feel more organized, consistent and structured, it also seems overdone, as e.g. `GRPCGatewayClientModule` afaics will only ever have one line like:
```golang
ecocredits.RegisterQueryHandlerClient(context.Background(), mux, ecocredits.NewQueryClient(clientCtx))
````

Reply:
The idea is not to expose the Server interfaces (I'm removing that interface), and the AppModule / ServerModule is not a server - it provides module information / descriptors.